File: integration_tests/java_client_tests_reactor_real/client-app/src/test/java/dev/graphlink/test/EnumsTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.enums.Priority;
import dev.graphlink.test.generated.enums.UserStatus;
import dev.graphlink.test.generated.types.GetUserResponse;
import dev.graphlink.test.generated.types.ListUsersByStatusResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class EnumsTest {

    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        client = newRealClient();
    }

    // ── Enum deserialization in response ──────────────────────────────────────

    @Nested
    class EnumDeserialization {

        @Test
        void status_deserializesToUserStatus_ACTIVE() {
            GetUserResponse res = client.queries.getUser("user-1").block();
            assertInstanceOf(UserStatus.class, res.getGetUser().getStatus());
            assertEquals(UserStatus.ACTIVE, res.getGetUser().getStatus());
        }

        @Test
        void status_deserializesToUserStatus_INACTIVE() {
            assertEquals(UserStatus.INACTIVE, client.queries.getUser("user-2").block().getGetUser().getStatus());
        }

        @Test
        void priority_deserializesToPriority_HIGH() {
            assertInstanceOf(Priority.class, client.queries.getUser("user-1").block().getGetUser().getPriority());
            assertEquals(Priority.HIGH, client.queries.getUser("user-1").block().getGetUser().getPriority());
        }

        @Test
        void priority_isNull_forBob() {
            assertNull(client.queries.getUser("user-2").block().getGetUser().getPriority());
        }
    }

    // ── Enum as query argument ────────────────────────────────────────────────

    @Nested
    class EnumAsQueryArgument {

        @Test
        void listUsersByStatus_ACTIVE_returnsOnlyActiveUsers() {
            ListUsersByStatusResponse res = client.queries.listUsersByStatus(UserStatus.ACTIVE).block();
            assertNotNull(res.getListUsersByStatus());
            assertFalse(res.getListUsersByStatus().isEmpty());
            res.getListUsersByStatus().forEach(u ->
                    assertEquals(UserStatus.ACTIVE, u.getStatus()));
        }

        @Test
        void listUsersByStatus_INACTIVE_returnsOnlyInactiveUsers() {
            ListUsersByStatusResponse res = client.queries.listUsersByStatus(UserStatus.INACTIVE).block();
            assertNotNull(res.getListUsersByStatus());
            assertFalse(res.getListUsersByStatus().isEmpty());
            res.getListUsersByStatus().forEach(u ->
                    assertEquals(UserStatus.INACTIVE, u.getStatus()));
        }
    }

    // ── All enum values round-trip ────────────────────────────────────────────

    @Nested
    class EnumRoundTrip {

        @Test
        void userStatus_ACTIVE_roundTrips() {
            assertEquals("ACTIVE", UserStatus.ACTIVE.toJson());
            assertEquals(UserStatus.ACTIVE, UserStatus.fromJson("ACTIVE"));
        }

        @Test
        void userStatus_INACTIVE_roundTrips() {
            assertEquals("INACTIVE", UserStatus.INACTIVE.toJson());
            assertEquals(UserStatus.INACTIVE, UserStatus.fromJson("INACTIVE"));
        }

        @Test
        void userStatus_SUSPENDED_roundTrips() {
            assertEquals("SUSPENDED", UserStatus.SUSPENDED.toJson());
            assertEquals(UserStatus.SUSPENDED, UserStatus.fromJson("SUSPENDED"));
        }

        @Test
        void priority_LOW_roundTrips() {
            assertEquals("LOW", Priority.LOW.toJson());
            assertEquals(Priority.LOW, Priority.fromJson("LOW"));
        }

        @Test
        void priority_CRITICAL_roundTrips() {
            assertEquals("CRITICAL", Priority.CRITICAL.toJson());
            assertEquals(Priority.CRITICAL, Priority.fromJson("CRITICAL"));
        }

        @Test
        void fromJson_null_returnsNull() {
            assertNull(UserStatus.fromJson(null));
            assertNull(Priority.fromJson(null));
        }
    }
}

File: integration_tests/kotlin_client_tests_real/client-app/src/test/kotlin/dev/graphlink/test/EnumsTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.Priority
import dev.graphlink.test.generated.enums.UserStatus
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.newRealClient

class EnumsTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = newRealClient()
    }

    // ── Enum deserialization in response ──────────────────────────────────────

    @Nested
    inner class EnumDeserialization {

        @Test
        fun status_deserializesToUserStatus_ACTIVE() = runTest {
            val res = client.queries.getUser("user-1")
            assertIs<UserStatus>(res.getUser.status)
            assertEquals(UserStatus.ACTIVE, res.getUser.status)
        }

        @Test
        fun status_deserializesToUserStatus_INACTIVE() = runTest {
            assertEquals(UserStatus.INACTIVE, client.queries.getUser("user-2").getUser.status)
        }

        @Test
        fun priority_deserializesToPriority_HIGH() = runTest {
            val priority = client.queries.getUser("user-1").getUser.priority
            assertIs<Priority>(priority)
            assertEquals(Priority.HIGH, priority)
        }

        @Test
        fun priority_isNull_whenAbsent() = runTest {
            assertNull(client.queries.getUser("user-2").getUser.priority)
        }
    }

    // ── Enum as query argument ────────────────────────────────────────────────

    @Nested
    inner class EnumAsQueryArgument {

        @Test
        fun response_containsDeserializedUsers() = runTest {
            val res = client.queries.listUsersByStatus(UserStatus.ACTIVE)
            assertNotNull(res.listUsersByStatus)
            assertTrue(res.listUsersByStatus.isNotEmpty())
            assertTrue(res.listUsersByStatus.all { it.status == UserStatus.ACTIVE })
        }
    }

    // ── All enum values round-trip ────────────────────────────────────────────

    @Nested
    inner class EnumRoundTrip {

        @Test
        fun userStatus_ACTIVE_roundTrips() {
            assertEquals("ACTIVE", UserStatus.ACTIVE.toJson())
            assertEquals(UserStatus.ACTIVE, UserStatus.fromJson("ACTIVE"))
        }

        @Test
        fun userStatus_INACTIVE_roundTrips() {
            assertEquals("INACTIVE", UserStatus.INACTIVE.toJson())
            assertEquals(UserStatus.INACTIVE, UserStatus.fromJson("INACTIVE"))
        }

        @Test
        fun userStatus_SUSPENDED_roundTrips() {
            assertEquals("SUSPENDED", UserStatus.SUSPENDED.toJson())
            assertEquals(UserStatus.SUSPENDED, UserStatus.fromJson("SUSPENDED"))
        }

        @Test
        fun priority_LOW_roundTrips() {
            assertEquals("LOW", Priority.LOW.toJson())
            assertEquals(Priority.LOW, Priority.fromJson("LOW"))
        }

        @Test
        fun priority_CRITICAL_roundTrips() {
            assertEquals("CRITICAL", Priority.CRITICAL.toJson())
            assertEquals(Priority.CRITICAL, Priority.fromJson("CRITICAL"))
        }

    }
}
